File: figaro-worker/src/worker/help-request.ts
/**
 * Worker-side help request handler for human-in-the-loop assistance.
 *
 * Subscribes per-request to the specific JetStream response subject
 * (figaro.help.{request_id}.response) via the HELP stream. JetStream
 * ensures responses survive NATS reconnections during long waits,
 * matching the Python supervisor's durable subscription approach.
 *
 * Ported from figaro-supervisor/src/figaro_supervisor/supervisor/client.py
 */

import type { NatsClient } from "../nats/client";
import { Subjects } from "../nats/subjects";
import type { HelpResponsePayload } from "../types";

export class HelpRequestHandler {
  private readonly client: NatsClient;

  constructor(client: NatsClient) {
    this.client = client;
  }

  /**
   * Request human help and wait for a response.
   *
   * Subscribes to the specific response subject via JetStream before
   * sending the request, ensuring no race conditions with delivery.
   * JetStream subscriptions survive NATS reconnections, so responses
   * are not lost during transient connection drops.
   *
   * @param taskId - The ID of the current task
   * @param questions - List of questions in AskUserQuestion format
   * @param context - Optional context to include with the request
   * @param timeoutSeconds - How long to wait for a response (default 30 min)
   * @returns A record mapping question text to answers, or null if timeout/error
   */
  async requestHelp(
    taskId: string,
    questions: Array<Record<string, unknown>>,
    context?: Record<string, unknown> | null,
    timeoutSeconds: number = 1800,
  ): Promise<Record<string, string> | null> {
    const requestId = crypto.randomUUID();
    const subject = Subjects.helpResponse(requestId);

    let settled = false;
    let settle: (value: HelpResponsePayload | null) => void;
    const responsePromise = new Promise<HelpResponsePayload | null>((res) => {
      settle = res;
    });

    // Subscribe via JetStream (HELP stream covers figaro.help.*.response)
    // before publishing the request to prevent race conditions.
    const sub = await this.client.subscribeJetStream(
      subject,
      (data: Record<string, unknown>) => {
        if (settled) return;
        if (data.request_id === requestId) {
          settled = true;
          settle(data as unknown as HelpResponsePayload);
        }
      },
    );

    try {
      // Publish the help request AFTER subscribing (race condition prevention)
      await this.client.publishHelpRequest(
        requestId,
        taskId,
        questions,
        timeoutSeconds,
      );

      // Race the response against a timeout
      const timeoutPromise = new Promise<null>((resolve) => {
        const timer = setTimeout(() => resolve(null), timeoutSeconds * 1000);
        // Allow the timer to not keep the process alive
        if (typeof timer === "object" && "unref" in timer) {
          timer.unref();
        }
      });

      const result = await Promise.race([responsePromise, timeoutPromise]);

      if (result === null) {
        return null;
      }

      // Check for errors in the response
      if (result.error) {
        return null;
      }

      return result.answers ?? null;
    } finally {
      settled = true;
      sub.unsubscribe();
    }
  }

  /**
   * Cancel all pending help requests.
   *
   * No-op: subscriptions are cleaned up in requestHelp's finally block.
   * @returns Count of cancelled requests (always 0).
   */
  cancelPendingRequests(): number {
    return 0;
  }
}

File: figaro-worker-legacy/src/figaro_worker/worker/help_request.py
"""Worker-side help request handler for human-in-the-loop assistance."""

import asyncio
import logging
from typing import Any
from uuid import uuid4

from figaro_nats import Subjects

from .client import NatsClient

logger = logging.getLogger(__name__)


class HelpRequestHandler:
    """
    Handles help requests from the worker to the orchestrator.

    Subscribes per-request to the specific Core NATS response subject
    (figaro.help.{request_id}.response) for reliable delivery. JetStream
    publish also delivers to Core NATS subscribers, so this works with
    the orchestrator's js_publish without needing a JetStream consumer.
    """

    def __init__(self, client: NatsClient) -> None:
        self._client = client
        self._pending_futures: dict[str, asyncio.Future[dict[str, str] | None]] = {}

    async def request_help(
        self,
        task_id: str,
        questions: list[dict[str, Any]],
        context: dict[str, Any] | None = None,
        timeout_seconds: int = 1800,
    ) -> dict[str, str] | None:
        """
        Request human help and wait for a response.

        Subscribes to the specific response subject via Core NATS before
        sending the request, ensuring no race conditions with delivery.

        Args:
            task_id: The ID of the current task
            questions: List of questions in AskUserQuestion format
            context: Optional context to include with the request
            timeout_seconds: How long to wait for a response (default 30 min)

        Returns:
            A dict mapping question text to answers, or None if timeout/error
        """
        request_id = str(uuid4())

        response_event = asyncio.Event()
        response_data: dict[str, Any] = {}

        async def _on_response(data: dict[str, Any]) -> None:
            if data.get("request_id") == request_id:
                response_data.update(data)
                response_event.set()

        # Subscribe to the specific response subject BEFORE sending the request
        sub = await self._client.conn.subscribe(
            Subjects.help_response(request_id),
            _on_response,
        )

        try:
            # Send the help request via NATS
            await self._client.publish_help_request(
                request_id=request_id,
                task_id=task_id,
                questions=questions,
                timeout_seconds=timeout_seconds,
            )

            logger.info(f"Sent help request {request_id} for task {task_id}, waiting for response...")

            try:
                await asyncio.wait_for(response_event.wait(), timeout=timeout_seconds)
            except asyncio.TimeoutError:
                logger.warning(f"Help request {request_id} timed out after {timeout_seconds}s")
                return None
            except asyncio.CancelledError:
                logger.info(f"Help request {request_id} was cancelled")
                return None

            # Check for errors
            error = response_data.get("error")
            if error:
                logger.warning(f"Help request {request_id} failed: {error}")
                return None

            answers = response_data.get("answers")
            logger.info(f"Received response for help request {request_id}")
            return answers
        finally:
            await sub.unsubscribe()

    def cancel_pending_requests(self) -> int:
        """Cancel all pending help requests. Returns count of cancelled requests."""
        # No-op: subscriptions are cleaned up in request_help's finally block
        return 0
